cache/blobs_linux.go

```diff
@@ -0,0 +1,398 @@
+			Source:  "overlay",
+			Options: []string{fmt.Sprintf("lowerdir=%s", strings.Join([]string{upperdir, emptyLower}, ":"))},
+		},
+	}
+	return mount.WithTempMount(ctx, lower, func(lowerRoot string) error {
+		return mount.WithTempMount(ctx, upperView, func(upperViewRoot string) error {
+			cw := archive.NewChangeWriter(w, upperViewRoot)
+			if err := overlayChanges(ctx, cw.HandleChange, upperdir, upperViewRoot, lowerRoot); err != nil {
+				if err2 := cw.Close(); err2 != nil {
+					return errors.Wrapf(err, "failed torecord upperdir changes (close error: %v)", err2)
+				}
+				return errors.Wrapf(err, "failed torecord upperdir changes")
+			}
+			return cw.Close()
+		})
+	})
+}
+
+// overlayChanges is continuty's `fs.Change`-like method but leverages overlayfs's
+// "upperdir" for computing the diff. "upperdirView" is overlayfs mounted view of
+// the upperdir that doesn't contain whiteouts. This is used for computing
+// changes under opaque directories.
+func overlayChanges(ctx context.Context, changeFn fs.ChangeFunc, upperdir, upperdirView, base string) error {
+	opaqueDirs := make(map[string]struct{})
+	return filepath.Walk(upperdir, func(path string, f os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Rebase path
+		path, err = filepath.Rel(upperdir, path)
+		if err != nil {
+			return err
+		}
+		path = filepath.Join(string(os.PathSeparator), path)
+
+		// Skip root
+		if path == string(os.PathSeparator) {
+			return nil
+		}
+
+		// Skip if this file is under an opaque directory that is already handled.
+		for opq := range opaqueDirs {
+			if strings.HasPrefix(path, opq) {
+				return nil
+			}
+		}
+
+		// Check if this is a deleted entry
+		isDelete, skip, err := checkDelete(upperdir, path, base, f)
+		if err != nil {
+			return err
+		} else if skip {
+			return nil
+		}
+
+		var kind fs.ChangeKind
+		var skipRecord bool
+		if isDelete {
+			// This is a deleted entry.
+			kind = fs.ChangeKindDelete
+			f = nil
+		} else if baseF, err := os.Lstat(filepath.Join(base, path)); err == nil {
+			// File exists in the base layer. Thus this is modified.
+			kind = fs.ChangeKindModify
+			// Avoid including directory that hasn't been modified. If /foo/bar/baz is modified,
+			// then /foo will apper here even if it's not been modified because it's the parent of bar.
+			if same, err := sameDir(baseF, f, filepath.Join(base, path), filepath.Join(upperdirView, path)); same {
+				skipRecord = true // Both are the same, don't record the change
+			} else if err != nil {
+				return err
+			}
+		} else if os.IsNotExist(err) {
+			// File doesn't exist in the base layer. Thus this is added.
+			kind = fs.ChangeKindAdd
+		} else if err != nil {
+			return err
+		}
+
+		if !skipRecord {
+			if err := changeFn(kind, path, f, nil); err != nil {
+				return err
+			}
+		}
+
+		if f != nil {
+			if isOpaque, err := checkOpaque(upperdir, path, base, f); err != nil {
+				return err
+			} else if isOpaque {
```

Comment:
nit: when `isOpaque` is true, could we return `filepath.SkipDir` as the error? Seems like it might be a simpler way to prevent checking files/dirs under an opaque dir than maintaining the `opaqueDirs` map and checking prefixes.

Reply:
Great idea! Fixed to use `filepath.SkipDir` instead of the map.